### PR TITLE
kinetis: Add DMA support to UART driver

### PR DESCRIPTION
**Pull request by gebart**
This is a refactoring of the UART driver on Kinetis platforms, including:
- Use `gpio_init_af` to select pin alternate function (introduced in #4830).
- Protect `uart_write` with a per device mutex
- Add support for up to 5 UART modules (was: 2)
- Add option to use DMA for UART TX

The DMA usage is similar to how the implementation works on STM32F4: Using a mutex to suspend the calling thread until the DMA ISR is called.

### CODE REVIEW

**Comment by gebart** [comment]
ping @haukepetersen

File: cpu/kinetis_common/include/periph_cpu.h [hunk 1]
```diff
@@ -166,6 +166,14 @@ typedef struct {
  */
 void gpio_init_port(gpio_t pin, uint32_t pcr);
 
+/**
+ * @brief Override uart_t
+ */
+#define HAVE_UART_T
+typedef uint8_t uart_t;
+#define UART_DEV(x)         (x)
+#define UART_UNDEF          0xff
```
Comment (haukepetersen) [inline cpu/kinetis_common/include/periph_cpu.h]: why overriding the default values here? At least the `UART_DEV(x)` macro is identical to the one in `uart.h`

File: cpu/kinetis_common/periph/uart.c [hunk 2]
```diff
@@ -64,32 +143,62 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
     config[uart].arg = arg;
 
-    /* enable receive interrupt */
+    /* enable interrupt source */
     switch (uart) {
 #if UART_0_EN
-
-        case UART_0:
+        case UART_DEV(0):
             NVIC_SetPriority(UART_0_IRQ_CHAN, UART_IRQ_PRIO);
```
Comment (haukepetersen) [inline cpu/kinetis_common/periph/uart.c]: no need to set the IRQ priority here anymore, it's done in the `cortexm_init()`

File: cpu/k60/include/cpu_conf.h [hunk 3]
```diff
@@ -97,6 +97,95 @@ extern "C"
...
+typedef enum {
+    DMA_ACCESS_SIZE_8BIT       = 0,
+    DMA_ACCESS_SIZE_16BIT      = 1,
+    DMA_ACCESS_SIZE_32BIT      = 2,
+    DMA_ACCESS_SIZE_RESERVED3  = 3,
+    DMA_ACCESS_SIZE_16BYTE     = 4,
+    DMA_ACCESS_SIZE_32BYTE     = 5,
+    DMA_ACCESS_SIZE_RESERVED6  = 6,
+    DMA_ACCESS_SIZE_RESERVED7  = 7,
+} dma_access_size_t;
```
Comment (haukepetersen) [inline cpu/k60/include/cpu_conf.h]: doesn't it make more sense to move this type to the kinetis_common code path?

**Comment by haukepetersen** [comment]
looks good in general, though I would prefer a complete UART remodeling to get rid of those #ifdef orgies, but I guess this might as well be done in a separate PR...

**Comment by haukepetersen** [comment]
@gebart: will you address the comments before tomorrow evening so we can get this into the release? Otherwise I would suggest to re-tag this PR for the 2016.07 release...

**Comment by gebart** [comment]
Are the DMA access size enum values the same for all kinetis CPUs? My guess is yes, since the DMA hardware block is probably close to identical between different models. I'll move the typedef for now and we can rework it if we discover a CPU which has different values.

File: cpu/kinetis_common/periph/uart.c (hunk 2 above)
Comment (gebart) [inline cpu/kinetis_common/periph/uart.c]: I think this should go in a separate PR which gets rid of the whole switch block

**Comment by gebart** [comment]
rebased, moved dma access size typedef and updated the code to follow renames of some symbols.

**Comment by gebart** [comment]
OK to squash?

**Comment by gebart** [comment]
rebased, added a DEVELHELP only DMA error handler which prints DMA debugging information to stdout to assist in finding configuration errors etc while developing.

File: cpu/k64f/include/cpu_conf.h [hunk 4]
```diff
@@ -71,6 +71,33 @@ extern "C"
...
+typedef enum {
+    DMA_ACCESS_SIZE_8BIT       = 0,
+    DMA_ACCESS_SIZE_16BIT      = 1,
+    DMA_ACCESS_SIZE_32BIT      = 2,
+    DMA_ACCESS_SIZE_RESERVED3  = 3,
+    DMA_ACCESS_SIZE_16BYTE     = 4,
+    DMA_ACCESS_SIZE_32BYTE     = 5,
+    DMA_ACCESS_SIZE_RESERVED6  = 6,
+    DMA_ACCESS_SIZE_RESERVED7  = 7,
+} dma_access_size_t;
```
Comment (haukepetersen) [inline cpu/k64f/include/cpu_conf.h]: this was moved to the common code path, right?

**Comment by haukepetersen** [comment]
There are two minor remarks (in cpu/k64f/include/cpu_conf.h and cpu/kinetis_common/include/periph_cpu.h). Please address them and squash.

File: cpu/kinetis_common/include/periph_cpu.h (hunk 1 above)
Comment (gebart) [inline cpu/kinetis_common/include/periph_cpu.h]: I honestly don't remember what was the reasoning behind it.. I will remove it for now, let's clean up the driver in a separate PR.

**Comment by gebart** [comment]
Rebased, squashed, addressed comments

**Comment by haukepetersen** [comment]
not working for `pba-d-01-kw2x`

**Comment by haukepetersen** [comment]
but might be unrelated to this PR?! -> #5223

**Comment by gebart** [comment]
Rebased, updated DMA driver to set some useful defaults on the DMA_CR register during init which I will need in another DMA application, but doesn't hurt to enable globally here..

@haukepetersen I don't understand what has gone wrong with the pba-d-01-kw2x. Does the board hang completely or does it simply not display anything on the console UART?

File: cpu/kinetis_common/periph/uart.c [hunk 5]
```diff
@@ -177,46 +297,126 @@ static int init_base(uart_t uart, uint32_t baudrate)
...
+            /* Set minor loop count */
+            DMA_DEV->TCD[channel].NBYTES_MLOFFNO = DMA_NBYTES_MLOFFNO_NBYTES(1);
+            /* Set major loop count */
+            /* The DMA engine can only handle up to 2**15 major loop iterations */
+            assert(len <= (DMA_CITER_ELINKNO_CITER_MASK >> DMA_CITER_ELINKNO_CITER_SHIFT));
+            DMA_DEV->TCD[channel].CITER_ELINKNO = DMA_CITER_ELINKNO_CITER(len);
+            DMA_DEV->TCD[channel].BITER_ELINKNO = DMA_BITER_ELINKNO_BITER(len);
+            /* Set control register */
+            /* Disable DMA requests after major loop finishes, interrupt on end of major loop */
+            DMA_DEV->TCD[channel].CSR = DMA_CSR_DREQ_MASK | DMA_CSR_INTMAJOR_MASK;
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/periph/uart.c]: @gebart Can this DMA specific code (L333 ... L351) be moved to cpu/kinetis_common/periph/dma.c ?

Comment (gebart) [inline cpu/kinetis_common/periph/uart.c]: I've been trying to come up with some way to do it but the DMA module is simply too flexible and any API solutions would cause significant overhead or reduce the functionality. For this UART implementation some kind of simplified API would still require at least the following parameters:
destination address, destination step size, destination access width, source address, source step size, source access width, bytes per chunk (minor loop), number of chunks (major loop).

For a more flexible solution (try for example to sample multiple ADC lines consecutively, similar to http://cache.nxp.com/files/32bit/doc/app_note/AN4590.pdf) where DMA channel linking and different offsets are required the API boils down to writing a setter function for each individual DMA TCD component register (except for BITER+CITER which must always match each other's values when configuring)...

**Comment by haukepetersen** [comment]
sorry for the delay: on the phytec board (running the default example), I simply get no output on the UART whatsoever.

File: cpu/kinetis_common/periph/uart.c [hunk 6]
```diff
@@ -64,85 +143,120 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
+        case UART_DEV(4):
+            NVIC_SetPriority(UART_4_IRQ_CHAN, UART_IRQ_PRIO);
+            NVIC_EnableIRQ(UART_4_IRQ_CHAN);
+            break;
+#endif
         default:
             return -2;
-            break;
+    }
+    KINETIS_UART *dev = _uart_base_ptrs[uart];
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/periph/uart.c]: BUG: uart can not be used for it because UART_0_DEV is UART2.

File: cpu/kinetis_common/periph/uart.c [hunk 7]
```diff
@@ -64,85 +143,120 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
+            gpio_init_port(UART_4_TX_GPIO, UART_4_TX_AF);
+            gpio_init_port(UART_4_RX_GPIO, UART_4_RX_AF);
+            clk = UART_4_CLK;
+            UART_4_CLKEN();
+            break;
+#endif
         default:
             return -1;
     }
+    KINETIS_UART *dev = _uart_base_ptrs[uart];
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/periph/uart.c]: BUG: also here, uart can not be used for it because UART_0_DEV is UART2.

**Comment by jfischer-phytec-iot** [comment]
^ I will write tonight a patch and test it.

**Comment by gebart** [comment]
I understand your comments as the problem is simply a configuration mistake
from my side (periph_conf.h).

**Comment by kYc0o** [comment]
Since it's an enhancement, I propose to postpone it, unless there is a chance to get it merged today...

**Comment by miri64** [comment]
Postponed due to Feature Freeze.

**Comment by OlegHahm** [comment]
Can we this in for the release?

**Comment by gebart** [comment]
Putting this one on hold for the time being. I'd like to focus on merging #5616 and #5688 first.

**Comment by miri64** [comment]
Both are merged now. You want to pick up work here again?

**Comment by haukepetersen** [comment]
@gebart is this PR still valid as #6868 remodeled the UART driver?

**Comment by gebart** [comment]
Abandoned for now, but the DMA stuff would be useful in the periph drivers.
